### PR TITLE
updates for TDP article for 2017-01-01.4.0

### PR DESCRIPTION
**Pull request by tamram**

### CODE REVIEW

**Comment by tamram** [comment]
Not ready for merge

File: CSharp/ArticleProjects/DotNetTutorial/Program.cs [hunk 1]
```diff
@@ -27,13 +27,13 @@ public class Program
         // These are used when constructing connection strings for the Batch and Storage client objects.
 
         // Batch account credentials
-        private const string BatchAccountName = "";
-        private const string BatchAccountKey  = "";
-        private const string BatchAccountUrl  = "";
+        private const string BatchAccountName = "tamrambatch";
+        private const string BatchAccountKey  = "a3HjJUjKtsuXwN7pshA7XdL3C6OAQQsof4+EML7n8IlyGdqJloGfk1qkbmfzWzdWafP+mu9jbjDwHB65UBEzZw==";
```
Comment (alfpark) [inline CSharp/ArticleProjects/DotNetTutorial/Program.cs]: Looks like some credential leakage happened. Can you rotate your keys?

File: CSharp/ArticleProjects/DotNetTutorial/Program.cs [hunk 2]
```diff
@@ -27,13 +27,13 @@ public class Program
...
+        private const string BatchAccountUrl  = "https://tamrambatch.westcentralus.batch.azure.com";
 
         // Storage account credentials
-        private const string StorageAccountName = "";
-        private const string StorageAccountKey  = "";
+        private const string StorageAccountName = "tamrambatchstorage";
+        private const string StorageAccountKey  = "PPBIRREmpThVmZ69urBX10vgN3QlYptajzL3kE+Ig4s0qfQNyLZy8RsKn9V+BDN6GynA6cteUWPNtVDMPz1IOA==";
```
Comment (alfpark) [inline CSharp/ArticleProjects/DotNetTutorial/Program.cs]: Rotate this as well.

File: CSharp/Common/AccountSettings.settings [hunk 3]
```diff
@@ -1,24 +1,24 @@
 ﻿<?xml version='1.0' encoding='utf-8'?>
-<SettingsFile xmlns="http://schemas.microsoft.com/VisualStudio/2004/01/settings" CurrentProfile="(Default)" GeneratedClassNamespace="Microsoft.Azure.Batch.Samples.Common" GeneratedClassName="Settings">
+<SettingsFile xmlns="http://schemas.microsoft.com/VisualStudio/2004/01/settings" CurrentProfile="(Default)" GeneratedClassNamespace="Microsoft.Azure.Batch.Samples.Common" GeneratedClassName="AccountSettings">
   <Profiles />
   <Settings>
     <Setting Name="BatchServiceUrl" Type="System.String" Scope="User">
-      <Value Profile="(Default)">https://(YourAccount).(region).batch.azure.com</Value>
```
Comment (alfpark) [inline CSharp/Common/AccountSettings.settings]: Revert all of these settings.

File: CSharp/Common/app.config [hunk 4]
```diff
@@ -1,38 +1,47 @@
...
-            <section name="Microsoft.Azure.Batch.Samples.Common.Settings" type="System.Configuration.ClientSettingsSection, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" allowExeDefinition="MachineToLocalUser" requirePermission="false" />
-        </sectionGroup>
-    </configSections>
+
     <userSettings>
-        <Microsoft.Azure.Batch.Samples.Common.Settings>
+        <Microsoft.Azure.Batch.Samples.Common.AccountSettings>
             <setting name="BatchServiceUrl" serializeAs="String">
-                <value>https://(YourAccount).(region).batch.azure.com</value>
+                <value>https://tamrambatch.westcentralus.batch.azure.com</value>
```
Comment (alfpark) [inline CSharp/Common/app.config]: Revert all of these settings

**Comment by tamram** [comment]
@alfpark, I will take these out and regenerate keys.

**Comment by tamram** [comment]
@alfpark , keys removed and regenerated.

File: CSharp/ArticleProjects/TaskDependencies/App.config [hunk 5]
```diff
@@ -7,7 +7,27 @@
     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
       <dependentAssembly>
         <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral" />
-        <bindingRedirect oldVersion="0.0.0.0-6.0.0.0" newVersion="6.0.0.0" />
+        <bindingRedirect oldVersion="0.0.0.0-9.0.0.0" newVersion="9.0.0.0" />
+      </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.Data.Services.Client" publicKeyToken="31bf3856ad364e35" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-5.6.4.0" newVersion="5.6.4.0" />
```
Comment (tamram) [inline CSharp/ArticleProjects/TaskDependencies/App.config]: I didn't change this explicitly, but I'll check the packages again.

File: CSharp/ArticleProjects/TaskDependencies/Program.cs [hunk 6]
```diff
@@ -58,105 +56,188 @@ public static void Main(string[] args)
...
+                        // Task 5 depends on a range of tasks, 1 through 3, and 'Flowers'
+                        new CloudTask("5", "cmd.exe /c echo 5")
+                        {
+                            DependsOn = new TaskDependencies(
+                                new[] { "Flowers" },
+                                new[] { new TaskIdRange(1, 3) })
+                        },
+
+                        // Task A is a parent task that exits with a non-zero exit code.
+                        new CloudTask("A", "cmd.exe /c echo A exit 1"),
```
Comment (tamram) [inline CSharp/ArticleProjects/TaskDependencies/Program.cs]: Okay, thanks. I need to look at more examples of command-line syntax as it's not something I've used in this way before.

File: CSharp/ArticleProjects/TaskDependencies/Program.cs [hunk 7]
```diff
@@ -58,105 +56,188 @@ public static void Main(string[] args)
...
+                        // Task C is a parent task that cannot run and so causes a scheduling error.
+                        new CloudTask("C", "NonExistentProgram.exe"),
+                        // Task D depends on task C.
+                        new CloudTask("D", "cmd.exe /c echo D")
+                        {
+                            DependsOn = TaskDependencies.OnId("C"),
+                            ExitConditions = new ExitConditions()
+                            {
+                                // When task C exits with a scheduling error, the downstream task D will not run.
+                                SchedulingError = new ExitOptions()
```
Comment (tamram) [inline CSharp/ArticleProjects/TaskDependencies/Program.cs]: Yes, that makes sense. The example you looked at previously showed setting both default and scheduling error behavior under the same task (so you read it properly on your phone). I'll change the example to do that again.

File: CSharp/ArticleProjects/TaskDependencies/Program.cs [hunk 8]
```diff
@@ -58,105 +56,188 @@ public static void Main(string[] args)
...
+                BatchClient batchClient = await BatchClient.OpenAsync(cred);
 
                 // Clean up the resources we've created in the Batch account
                 Console.Write("Delete job? [yes] no: ");
                 string response = Console.ReadLine().ToLower();
                 if (response != "n" && response != "no")
                 {
-                    await batchClient.JobOperations.DeleteJobAsync(job.Id);
+                    await batchClient.JobOperations.DeleteJobAsync(jobId);
+                    Console.WriteLine("Job '{0}' deleted.", jobId);
```
Comment (tamram) [inline CSharp/ArticleProjects/TaskDependencies/Program.cs]: Okay, good to know. This was in the example already; I just changed it to use the jobId variable and put it in a Finally block. As I said in my email, I think I'm going to need to dig deeper here than I have time for right now, so perhaps we can review the whole sample together to make sure what's there all makes sense.

File: CSharp/ArticleProjects/TaskDependencies/TaskDependencies.csproj [hunk 9]
```diff
@@ -38,36 +38,36 @@
...
-    <Reference Include="Microsoft.Azure.KeyVault.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>packages\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
+    <Reference Include="Microsoft.Azure.KeyVault.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>packages\Microsoft.Azure.KeyVault.Core.2.0.4\lib\net45\Microsoft.Azure.KeyVault.Core.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Data.Edm, Version=5.7.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>packages\Microsoft.Data.Edm.5.7.0\lib\net40\Microsoft.Data.Edm.dll</HintPath>
+    <Reference Include="Microsoft.Data.Edm, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>packages\Microsoft.Data.Edm.5.8.2\lib\net40\Microsoft.Data.Edm.dll</HintPath>
```
Comment (tamram) [inline CSharp/ArticleProjects/TaskDependencies/TaskDependencies.csproj]: I'll check the packages again. Not sure what's going on there.

File: CSharp/Common/AccountSettings.settings [hunk 10]
```diff
@@ -1,24 +1,24 @@
 ﻿<?xml version='1.0' encoding='utf-8'?>
-<SettingsFile xmlns="http://schemas.microsoft.com/VisualStudio/2004/01/settings" CurrentProfile="(Default)" GeneratedClassNamespace="Microsoft.Azure.Batch.Samples.Common" GeneratedClassName="Settings">
+<SettingsFile xmlns="http://schemas.microsoft.com/VisualStudio/2004/01/settings" CurrentProfile="(Default)" GeneratedClassNamespace="Microsoft.Azure.Batch.Samples.Common" GeneratedClassName="AccountSettings">
```
Comment (tamram) [inline CSharp/Common/AccountSettings.settings]: I don't know - I didn't change this explicitly. I used the Settings manager to update the settings.

File: CSharp/ArticleProjects/TaskDependencies/Program.cs [hunk 11]
```diff
@@ -58,105 +56,188 @@ public static void Main(string[] args)
...
+
+                        // Task A is a parent task that exits with a non-zero exit code.
+                        new CloudTask("A", "cmd.exe /c echo A exit 1"),
+                        // Task B depends on task A
+                        new CloudTask("B", "cmd.exe /c echo B")
+                        {
+                            DependsOn = TaskDependencies.OnId("A"),
+                            ExitConditions = new ExitConditions()
+                            {
+                                // If task A exits with a non-zero exit code, run downstream task B.
```
Comment (itowlson) [inline CSharp/ArticleProjects/TaskDependencies/Program.cs]: To have an exit option for non-zero exit codes *only*, you would need to either:

1. Use ExitCodeRanges to specify a range that takes in all possible codes, e.g. 1 to 2147483647

2. Set Default to the desired 'all non-zero exit codes' behaviour and have an explicit SchedulingError clause to override it for the scheduling error case

**Comment by matthchr** [comment]
@tamram What's the status on these changes -- it looks like there are still a few comments left?

**Comment by tamram** [comment]
@matthchr - I am going to close this PR and start over on these changes, so as to avoid the package reference issues. I had really only opened it so that I could ask Ivan some questions about the code, but I should have communicated that more clearly.

**Pull request closed without merging**